Tests/test_file.py

```diff
@@ -787,6 +787,10 @@ def test_open_abplus(self):
         with open(self.temp_file, "ab+") as f:
             f.write(b"abc")
             f.seek(0)
-            self.assertEqual(f.read(), b"abc")
+            self.assertEqual(f.read(2), b"ab")
```

Comment:
I guess my brain doesn't understand `ab+` mode. :smile: Played around trying to make sense of it, but just ended up finding a behaviour that diverged from CPython...
```py
with open(self.temp_file, "ab+") as f:
    f.write(b"abcdef")
    f.seek(0)
    f.read(2)
    f.write(b"z")
    f.seek(0)
    self.assertEqual(f.read(), b"abzdef")
````

Reply:
I also do not fully understand the `ab+` mode, so I was playing around with some samples and tests to get a better grip on it; this test being one of them. So this change actually does not belong to this PR, kind of sneaked through from those experiments into the commit. It doesn't hurt though, the more test coverage the better, so I intend to leave it in, but also intend add more `ab+` testing later on.

Another thing I don't understand is why `ab+` is implemented using separate read and write streams. The dual-stream case does not play well with having one genuine file descriptor per file, but more worrisome is that I do not see any actual need for dual-streams at all. This worries me because maybe I am overlooking something essential that led the original programmer to use dual streams, hence the extra tests to try to catch something that doesn't fit in my current mental model of `ab+`. If I don't find anything, the next PR will probably get rid of dual streams.

Reply:
> I guess my brain doesn't understand `ab+` mode. 😄 Played around trying to make sense of it, but just ended up finding a behaviour that diverged from CPython...

It looks to me that you have found a bug in CPython. If you write four "z" rather than one, it is being correctly appended at the end. I was able to reproduce the strange behaviour you have found on macOS with various CPython versions, up to 3.12.0. Unrelated to this, but I've found [open issues around `ab+` in CPython](https://github.com/python/cpython/issues/86772), so it does not surprise me if there are more.

I have read the CPython code dealing with `ab+` and it simply adds `O_APPEND | O_CREAT` for file opening (in addition to the other usual flags) plus an initial seek to the end in case the first operation is read iso write. Further it relays on the OS to support it, which, according to POSIX documentation, should perform seek to end before each write atomically (apparently, non-atomically on Windows). However, using `pwrite`, it is still possible to write inside a file that is `O_APPEND`, so maybe this is what is happening for writes less than 4 bytes. Or maybe no OS call is done but CPython just updates some internal buffer as an optimization.

In any case, I propose to disregard all discrepancies and simply always write to the end, using one read/write stream.